src/java/picard/reference/NonNFastaSize.java

```diff
@@ -0,0 +1,103 @@
+        usageShort = NonNFastaSize.USAGE_SUMMARY,
+        programGroup = Fasta.class
+)
+public class NonNFastaSize extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Counts the number of non-N bases in a fasta file.";
+    static final String USAGE_DETAILS = "This tool takes any FASTA-formatted file and counts the number of non-N bases in it." +
+            "Note that it requires that the fasta file have associated index (.fai) and dictionary (.dict) files.<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar NonNFastaSize \\<br />" +
+            "      I=input_sequence.fasta \\<br />" +
+            "      O=count.txt" +
+            "</pre>" +
+            "<hr />"
+            ;
+    @Option(shortName= StandardOptionDefinitions.INPUT_SHORT_NAME, doc="The input FASTA file.")
+    public File INPUT;
+
+    @Option(shortName= StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc="The output file in which to record the count.")
+    public File OUTPUT;
+
+    @Option(shortName = "INTERVALS", doc = "An interval list file that contains the locations of the positions to assess.  If not provided, the entire reference will be used", optional = true)
+    public File INTERVALS = null;
+
+    public static void main(final String[] args) {
+        new NonNFastaSize().instanceMainWithExit(args);
+    }
+
+    @Override
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFileIsWritable(OUTPUT);
+
+        // set up the reference and a mask so that we only count the positions requested by the user
+        final ReferenceSequenceFile ref = ReferenceSequenceFileFactory.getReferenceSequenceFile(INPUT);
+        final ReferenceSequenceMask referenceSequenceMask;
+        if (INTERVALS != null) {
+            IOUtil.assertFileIsReadable(INTERVALS);
+            final IntervalList intervalList = IntervalList.fromFile(INTERVALS);
+            referenceSequenceMask = new IntervalListReferenceSequenceMask(intervalList);
+        } else {
+            final SAMFileHeader header = new SAMFileHeader();
+            header.setSequenceDictionary(ref.getSequenceDictionary());
+            referenceSequenceMask = new WholeGenomeReferenceSequenceMask(header);
+        }
+
+        long nonNbases = 0L;
+
+        for (final SAMSequenceRecord rec : ref.getSequenceDictionary().getSequences()) {
+            // pull out the contig and set up the bases
+            final ReferenceSequence sequence = ref.getSequence(rec.getSequenceName());
+            final byte[] bases = sequence.getBases();
+            StringUtil.toUpperCase(bases);
+
+            for (int i = 0; i < bases.length; i++) {
+                // only investigate this position if it's within our mask
+                if (referenceSequenceMask.get(sequence.getContigIndex(), i+1)) {
+                    nonNbases += bases[i] == 'N' ? 0 : 1;
```

Comment:
use `SequenceUtil.N` instead of `N` (perhaps overkill).  What about IUPAC bases, does that matter for your purposes?  If so, you could use `!SequenceUtil.isValidBase`.